keys.go

```diff
@@ -97,52 +95,46 @@ func (keys ServerKeys) PublicKey(keyID KeyID, atTS Timestamp) []byte {
 
 // FetchKeysDirect fetches the matrix keys directly from the given address.
 // Optionally sets a SNI header if ``sni`` is not empty.
-// Returns the server keys and the state of the TLS connection used to retrieve them.
-func FetchKeysDirect(serverName ServerName, addr, sni string) (*ServerKeys, *tls.ConnectionState, error) {
-	// Create a TLS connection.
-	tcpconn, err := net.Dial("tcp", addr)
-	if err != nil {
-		return nil, nil, err
+// Optionally sets a timeout to the HTTP client if ``timeout`` isn't 0.
+// Note that this function doesn't check the validity of the certificate(s)
+// served by the server.
+// Returns the server keys and the state of the TLS connection used to retrieve
+// them.
+func FetchKeysDirect(addr, sni string, timeout time.Duration) (*ServerKeys, *tls.ConnectionState, error) {
+	cli := http.Client{
+		Timeout: timeout, // A 0 timeout means no timeout.
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				ServerName: sni,
+				// TODO: Remove this once Synapse 1.0 is out.
+				InsecureSkipVerify: true, // nolint: gas
+			},
+		},
 	}
-	defer tcpconn.Close() // nolint: errcheck
-	tlsconn := tls.Client(tcpconn, &tls.Config{
-		ServerName: sni,
-
-		// This must be specified even though the TLS library will ignore it.
-		InsecureSkipVerify: true, // nolint: gas
-	})
-	if err = tlsconn.Handshake(); err != nil {
-		return nil, nil, err
-	}
-	connectionState := tlsconn.ConnectionState()
 
-	// Write a GET /_matrix/key/v2/server down the connection.
-	requestURL := "matrix://" + string(serverName) + "/_matrix/key/v2/server"
+	// Create a GET /_matrix/key/v2/server request.
+	requestURL := "https://" + addr + "/_matrix/key/v2/server"
```

Comment:
are you sure this sets the `Host` header correctly? I think you need to specify the original server name here. See how the federation client does it.

Reply:
Good catch, fixed in https://github.com/matrix-org/gomatrixserverlib/pull/126/commits/fd69519046eba39779262e1f178dd292174a116f